mypy/subtypes.py

```diff
@@ -303,6 +331,227 @@ def visit_type_type(self, left: TypeType) -> bool:
+            item = (member,
+                    get_member_flags(member, left.type),
+                    get_member_flags(member, right.type))
+            all_flags.append(item)
+    return all_flags
+
+
+def get_member_flags(name: str, info: TypeInfo) -> Set[int]:
+    """Detect whether a member 'name' is settable, whether it is an
+    instance or class variable, and whether it is class or static method.
+    """
+    method = info.get_method(name)
+    setattr_meth = info.get_method('__setattr__')
+    if method:
+        # this could be settable property
+        if method.is_property:
+            assert isinstance(method, OverloadedFuncDef)
+            dec = method.items[0]
+            assert isinstance(dec, Decorator)
+            if dec.var.is_settable_property or setattr_meth:
+                return {IS_SETTABLE}
+        return set()
+    node = info.get(name)
+    if not node:
+        if setattr_meth:
+            return {IS_SETTABLE}
+        return set()
+    v = node.node
+    if isinstance(v, Decorator):
+        if v.var.is_staticmethod or v.var.is_classmethod:
+            return {IS_CLASS_OR_STATIC}
+    # just a variable
+    if isinstance(v, Var):
+        flags = {IS_SETTABLE}
+        if v.is_classvar:
+            flags.add(IS_CLASSVAR)
+        return flags
+    return set()
+
+
+def find_var_type(var: Var, itype: Instance, subtype: Type) -> Type:
+    """Find type of a variable 'var' (maybe also a decorated method).
+    Apply type arguments from 'itype', and bind 'self' to 'subtype'.
+    """
+    from mypy.checkmember import bind_self
+    itype = map_instance_to_supertype(itype, var.info)
+    typ = var.type
+    if typ is None:
+        return AnyType()
+    typ = expand_type_by_instance(typ, itype)
+    # We don't need to bind 'self' for static methods, since there is no 'self'.
+    if isinstance(typ, FunctionLike) and not var.is_staticmethod:
+        signature = bind_self(typ, subtype)
+        assert isinstance(signature, CallableType)
+        if var.is_property:
+            return signature.ret_type
+        return signature
+    return typ
+
+
+def map_method(method: FuncBase, itype: Instance, subtype: Type) -> Type:
+    """Map 'method' to the base where it was defined. Apply type arguments
+    from 'itype', and bind 'self' type to 'subtype'.
+    This function should be used only for non-decorated methods. Decorated
+    methods (including @staticmethod and @property) are treated
+    by 'find_var_type'.
+    """
+    from mypy.checkmember import bind_self
+    signature = function_type(method, fallback=Instance(itype.type.mro[-1], []))
+    signature = bind_self(signature, subtype)
+    itype = map_instance_to_supertype(itype, method.info)
+    return expand_type_by_instance(signature, itype)
+
+
+def get_missing_members(left: Instance, right: Instance) -> List[str]:
+    """Find all protocol members of 'right' that are not implemented
+    (i.e. completely missing) in 'left'.
```

Comment:
Mention in docstring for error reporting helpers that they are error reporting helpers. (Also group them together.)